client/client.go

```diff
@@ -100,99 +74,114 @@ func NewProtocolClient(endpoint string, schemaType pb.SchemaType,
+		id:       rand.Int31(),
+		pbClient: client,
+		conn:     conn,
+		stream:   stream,
 	}
 
-	logger.Infof("NewProtocol client spawned (%v)", protocolClient.id)
-	return &protocolClient, nil
+	logger.Infof("New GenericClient spawned (%v)", genClient.id)
+	return &genClient, nil
 }
 
-func (c *Client) send(msg *pb.Message) error {
-	if err := c.stream.Send(msg); err != nil {
-		return fmt.Errorf("[Client %v] Error sending message: %v", c.id, err)
-	}
-	logger.Infof("[Client %v] Successfully sent request:", c.id, msg)
+// ProtocolParams is a map containing values required to bootstrap
+// the chosen protocol.
+type ProtocolParams map[string]*big.Int
 
-	return nil
+func getConnection(serverEndpoint string) (*grpc.ClientConn, error) {
+	logger.Debug("Getting the connection")
+	timeoutSec := config.LoadTimeout()
+	dialOptions := []grpc.DialOption{
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithTimeout(time.Duration(timeoutSec) * time.Second),
+	}
+	conn, err := grpc.Dial(serverEndpoint, dialOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("Could not connect to server %v (%v)", serverEndpoint, err)
+	}
+	return conn, nil
 }
 
-func (c *Client) receive() (*pb.Message, error) {
-	resp, err := c.stream.Recv()
-	if err == io.EOF {
-		return nil, fmt.Errorf("[Client %v] EOF error", c.id)
-	} else if err != nil {
-		return nil, fmt.Errorf("[Client %v] An error ocurred: %v", c.id, err)
+func getStream(client pb.ProtocolClient) (pb.Protocol_RunClient, error) {
+	logger.Debug("Getting the stream")
+	stream, err := client.Run(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("Error creating the stream: %v", err)
 	}
-	logger.Infof("[Client %v] Received response from the stream: %v", c.id, resp)
-	return resp, nil
+	return stream, nil
 }
 
-// close closes the communication stream and connection to the server.
-func (c *Client) close() error {
-	if err := c.stream.CloseSend(); err != nil {
-		return fmt.Errorf("[Client %v] Error closing stream: %v", c.id, err)
+func validateSchema(schemaType pb.SchemaType, schemaVariant pb.SchemaVariant) error {
+	if _, success := pb.SchemaType_name[int32(schemaType)]; !success {
+		return fmt.Errorf("Invalid SchemaType: %v", schemaType)
 	}
-	if err := c.conn.Close(); err != nil {
-		return fmt.Errorf("[Client %v] Error closing connection: %v", c.id, err)
+	if _, success := pb.SchemaVariant_name[int32(schemaVariant)]; !success {
+		return fmt.Errorf("Invalid SchemaVariant: %v", schemaVariant)
 	}
 	return nil
 }
 
-func (c *Client) ExecuteProtocol(params ProtocolParams) {
-	schemaType := pb.SchemaType_name[int32(c.schema)]
-	schemaVariant := pb.SchemaVariant_name[int32(c.variant)]
-	variant := common.ToProtocolType(c.variant)
+func NewClient(endpoint string) *Client {
+	return &Client{
+		endpoint: endpoint,
+	}
+}
 
-	logger.Infof("Starting client [%v] %v (%v)", c.id, schemaType, schemaVariant)
+func (c *Client) ExecuteProtocol(schema pb.SchemaType, variant pb.SchemaVariant,
```

Comment:
You're absolutely right. I moved this logic to CLI (`emmy.go`) and removed the `ExecuteProtocol` function from `client.go` altogether. In addition, now that the user of the library will be directly instantiating the desired protocol client, this removes the need for `map[string]*big.Int protocolParams`  in `client.go` (which is something that I wanted to get rid of as soon as possible anyway :+1:).